### PR TITLE
"azurerm_role_assignment" supports property "delegated_managed_identity_resource_id"

### LINKED ISSUE #9348: Support for delegatedManagedIdentityResourceId to azurerm_role_definition resource
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Enable the Azure Lighthouse scenario described in this article with Terraform: [Deploy a policy that can be remediated within a delegated subscription](https://docs.microsoft.com/en-us/azure/lighthouse/how-to/deploy-policy-remediation)

In order to accomplish this, a new property has been added to role definitions in 2019-04-01-preview API. This property "delegatedManagedIdentityResourceId" is required for assigning roles in Azure Lighthouse. Update the [azurerm_role_assignment resource](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/role_assignment) to include the new property.

### New or Affected Resource(s)

* azurerm_role_assignment

### Potential Terraform Configuration

```hcl
resource "azurerm_role_assignment" "example" {
  scope                             = data.azurerm_subscription.primary.id
  role_definition_name              = "Reader"
  principal_id                      = data.azurerm_client_config.example.object_id
  delegated_managed_identity_id     = azurerm_user_assigned_identity.example.id
}
```

### PR DESCRIPTION
**Pull request by njuCZ**
fix #9348

docs: https://docs.microsoft.com/en-us/azure/lighthouse/how-to/deploy-policy-remediation

this property does not mean the managed identity resource id, but the resource which contains a managed identity.
It's used for a cross tenant scenario.  assume there are two tenants and two service principals in each tenant:

| Tenant| subscription | User|
| --- | --- | --- |
|tenantA|subscriptionA|userA|
|tenantB|subscriptionB|userB|

By this new property, userA in tenantA could do role assignment in subscriptionB.

specific test step:
1. **userB** should delegate subscriptionB to userA by using lighthouse RP. **userB** could do it in terraform in following script
```
provider "azurerm" {
  features {}
}

variable "managing_tenant_id" {
  type = string
  // default = tenantA
}

variable "managing_principal_id" {
  type = string
  // default = userA
}

data "azurerm_role_definition" "user_access_administrator" {
  role_definition_id = "18d7d88d-d35e-4fb5-a5c3-7773c20a72d9"
}

data "azurerm_role_definition" "contributor" {
  role_definition_id = "b24988ac-6180-42a0-ab88-20f7382dd24c"
}

data "azurerm_subscription" "test" {}

resource "azurerm_lighthouse_definition" "test" {
  name                     = "acctest-delegate"
  description              = "Acceptance Test Lighthouse Definition"
  managing_tenant_id       = var.managing_tenant_id
  scope                    = data.azurerm_subscription.test.id

  authorization {
    principal_id           = var.managing_principal_id
    role_definition_id     = data.azurerm_role_definition.user_access_administrator.role_definition_id
    delegated_role_definition_ids = [
      data.azurerm_role_definition.contributor.role_definition_id,
    ]
  }

  authorization {
    principal_id           = var.managing_principal_id
    role_definition_id     = data.azurerm_role_definition.contributor.role_definition_id
  }
}

resource "azurerm_lighthouse_assignment" "test" {
  scope                    = data.azurerm_subscription.test.id
  lighthouse_definition_id = azurerm_lighthouse_definition.test.id
}
```
2. **userA** should switch to use **subscriptionB**
3. **userA** execute the acctest

notes, there is an api limitation: 
for cross tenant role assignment, we should pass the tenantId in the request, otherwise it could not read the resource.
But in normal cases, we could not pass the tenantId property, otherwise it will report error.

**changes in this PR**:
to fix the above limitation, in normal case, I don't change the id format. For cross tenant scenario, I append the tenantId into the id. This is not breaking change.

### CODE REVIEW

**Review by katbyte — APPROVED** [review]
Thanks @njuCZ - LGTM 👍

File: azurerm/internal/services/authorization/role_assignment_resource.go [hunk 1]
```diff
@@ -197,19 +222,19 @@ func resourceArmRoleAssignmentCreate(d *pluginsdk.ResourceData, meta interface{}
 		properties.RoleAssignmentProperties.PrincipalType = authorization.ServicePrincipal
 	}
 
-	if err := pluginsdk.Retry(d.Timeout(pluginsdk.TimeoutCreate), retryRoleAssignmentsClient(d, scope, name, properties, meta)); err != nil {
+	if err := pluginsdk.Retry(d.Timeout(schema.TimeoutCreate), retryRoleAssignmentsClient(d, scope, name, properties, meta, tenantId)); err != nil {
```
Comment (njuCZ) [inline azurerm/internal/services/authorization/role_assignment_resource.go:225]: this should be `pluginsdk.TimeoutCreate`, I will submit a PR to fix it soon

Comment (katbyte) [reply azurerm/internal/services/authorization/role_assignment_resource.go:225]: ty i was trying to resolve it via github - didn't go as well as i'd hoped 😓

**Comment by w0ut0** [comment]
Is it possible this causes https://github.com/terraform-providers/terraform-provider-azurerm/issues/12060?

**Comment by HarleyB123** [comment]
Hey, I think @w0ut0 is right, see my findings on that issue [here](https://github.com/terraform-providers/terraform-provider-azurerm/issues/12060#issuecomment-854838040).

@njuCZ I'm not sure how all the pieces fit together - hoping you might be able to see where it's going wrong 😄

**Comment by aristosvo** [comment]
The problem seems to be surfacing when the scope is on resource level instead of subscription, resource group etc.

Unfortunately this is not a scenario covered in the Acceptance Tests as far as I can see.

**Comment by njuCZ** [comment]
@w0ut0, @HarleyB123, @aristosvo sorry for causing the problem. I am glad to see that PR #12076 could fix it.

**Comment by etaham** [comment]
Can this be patched? It is causing state corruption for existing role assignments.

**Pull request merged**